receiver/signalfxreceiver/signalfxv2_to_metricdata.go

```diff
@@ -0,0 +1,237 @@
+
+	// Combine metric type with the actual data point type
+	sfxMetricType := sfxDataPoint.GetMetricType()
+	sfxDatum := sfxDataPoint.Value
+	if sfxDatum == nil {
+		return metricspb.MetricDescriptor_UNSPECIFIED, errSFxNilDatum
+	}
+
+	switch sfxMetricType {
+	case sfxpb.MetricType_GAUGE:
+		// Numerical: Periodic, instantaneous measurement of some state.
+		descType = metricspb.MetricDescriptor_GAUGE_DOUBLE
+		if sfxDatum.IntValue != nil {
+			descType = metricspb.MetricDescriptor_GAUGE_INT64
+		}
+
+	case sfxpb.MetricType_COUNTER:
+		// Numerical: Count of occurrences. Generally non-negative integers.
+		fallthrough
+
+	case sfxpb.MetricType_CUMULATIVE_COUNTER:
+		// Tracks a value that increases over time, where only the difference is important.
+		descType = metricspb.MetricDescriptor_CUMULATIVE_DOUBLE
+		if sfxDatum.IntValue != nil {
+			descType = metricspb.MetricDescriptor_CUMULATIVE_INT64
+		}
+
+	case sfxpb.MetricType_ENUM:
+		// String: Used for non-continuous quantities (that is, measurements where there is a fixed
+		// set of meaningful values). This is essentially a special case of gauge.
+		// Attempt to treat it as a numeric gauge.
+		descType = metricspb.MetricDescriptor_GAUGE_DOUBLE
+
+	default:
+		err = fmt.Errorf("unknown data-point type (%d)", sfxMetricType)
+	}
+
+	return descType, err
+}
+
+func buildPoint(
+	sfxDataPoint *sfxpb.DataPoint,
+	expectedMetricType metricspb.MetricDescriptor_Type,
+) (*metricspb.Point, error) {
+	if sfxDataPoint.Value == nil {
+		return nil, errSFxNilDatum
+	}
+
+	p := &metricspb.Point{
+		Timestamp: convertTimestamp(sfxDataPoint.GetTimestamp()),
+	}
+
+	switch {
+	case sfxDataPoint.Value.IntValue != nil:
+		mismatch := expectedMetricType != metricspb.MetricDescriptor_CUMULATIVE_INT64 &&
+			expectedMetricType != metricspb.MetricDescriptor_GAUGE_INT64
+		if mismatch {
+			return nil, errSFxUnexpectedInt64DatumType
+		}
+		p.Value = &metricspb.Point_Int64Value{Int64Value: *sfxDataPoint.Value.IntValue}
+
+	case sfxDataPoint.Value.DoubleValue != nil:
+		mismatch := expectedMetricType != metricspb.MetricDescriptor_CUMULATIVE_DOUBLE &&
+			expectedMetricType != metricspb.MetricDescriptor_GAUGE_DOUBLE
+		if mismatch {
+			return nil, errSFxUnexpectedFloat64DatumType
+		}
+		p.Value = &metricspb.Point_DoubleValue{DoubleValue: *sfxDataPoint.Value.DoubleValue}
+
+	case sfxDataPoint.Value.StrValue != nil:
+		if expectedMetricType != metricspb.MetricDescriptor_GAUGE_DOUBLE {
+			return nil, errSFxUnexpectedStringDatumType
+		}
+		dbl, err := strconv.ParseFloat(*sfxDataPoint.Value.StrValue, 64)
+		if err != nil {
+			return nil, errSFxStringDatumNotNumber
+		}
+		p.Value = &metricspb.Point_DoubleValue{DoubleValue: dbl}
+
+	default:
+		return nil, errSFxNoDatumValue
+	}
+
+	return p, nil
+}
+
+func convertTimestamp(msec int64) *timestamp.Timestamp {
+	if msec == 0 {
+		return nil
+	}
+
+	ts := &timestamp.Timestamp{
+		Seconds: msec / 1e3,
+		Nanos:   int32(msec%1e3) * 1e6,
+	}
```

Comment:
Seeing this again and again. We should add it as a time util func somewhere after this PR is merged.